### ISSUE TITLE
Side effects of iron curtain power on autotargeting are too strong

### ISSUE THREAD

**Issue by deleted-user-1** [OpenRA/OpenRA#19008]
## Motivation

Basically a request to the gameplay-maintainers and competitive players to consider this for the RA mod:

Firstly, this is a "I don't like the current IC-mechanic" and not a "IC is unbalanced" type of request. I have no idea if it is balanced in 1v1.

**What I think is bad about the effects of the iron curtain power is that when the opponent casts his IC spell to some of his units in engagements, my units become useless against his other, vulnernable units .**

*This itself is not the real problem for me. In StarCraft: Broodwar, Zerg can cast the Dark-Swarm spell over their units (a medium-sized orange cloud) which is greatly effective against the Terran faction, as Terran has only units with ranged attacks and enemies under Dark-Swarm basically don't take damage from ranged attacks (with some exceptions). Although making the Zerg units not invulnerable, Terran usually has to run away. So one of the top priorities for the Terran is to kill the spellcaster unit (the defiler) with own spell casters that Zerg often can not defend against (irradiate it). The DS spell is medium range so Zerg has to bring its defiler to the fight. DS does not effect ranged-attacks against buildings. It's considered the most powerful spell in the game, often game-ending for Terran, can be used offensively and defensively but I think it's great.*

I consider the current effect of the iron curtain power bad for gameplay in RA for the following reasons:
- Conceptually: The iron curtain is meant to make a few units invulnerable, signalized by them being highlighted in red. It's not causing an orange cloud that affects all units in the area. But in current RA the power of the iron curtain is not the invulnerability of some units (as it should be), but the inability of the opponent to tell his units to not fire at the invulnerable units.
- Side-effects are too strong: Casting the IC power on some units in an engagement means it affects all your surrounding units positively while affecting all the surrounding enemy units negatively.
- Too easy: There is no skill involved in casting the iron curtain power in an engagement at one of your tanks, moving it into the enemy's much stronger army and consequently win by a-moving the rest of your inferior blob in that direction.
- Losing to the computer instead of the enemy's tactic: Engagements in RA typically mean to position big blobs of units and then let them autotarget the enemy blob. Yes, there is direct targeting involved but it is fundamentally based on attack-moving and statically autotargeting. This is because the backbone of armies in RA - infantry - are present in too great numbers and die too fast. Autotargeting works well and is better than you could ever target manually in the typical engagements of blobs in RA.
- Turns the exisiting defenders advantage into a "Nope, you can't attack here": If the iron curtain power is used on base defences it often means that you will lose an entire army when doing a big attack. It already is really hard to attack a base as defences are usually there in big numbers, have greater range than you and are rebuilt super fast. When your army suddenly stops dealing damage in such a position because a big chunk goes to the invulnerable tesla coil, you lose everything. Consequently, you back off when only seeing the iron curtain. In late-game-teamgames (on maps that theoretically could be more dynamic) this often means waiting for nukes or chronoshifting MCVs inside the enemy base and tower-push from inside.

What I would like to see more is:
- Well actually I hate this too but invulnerable flaks that destroy allied tech units in seconds is a good example for utilizing the iron curtain power (having invulnerable units). Takes a bit of micro and direct targeting to be effective.
- Better: Invulnerable V2s that snipe key structures and get back home. Takes some preparation and good execution opposed to the "one-click-engagement-winner" use case.
- Also good: Quickly using it on key structures before the V2/nuke hits or saving that Conyard at 1% health.

## Proposed solution

A couple of ways could be explored:
- Tell `AutoTargetPriority` that actors with iron curtain condition are `InvalidTargets`. Means that units will never attack these units. Could look confusing when there are *only* invulnerable units and nothing is attacking them.
- Tell `AutoTargetPriority` that actors with iron curtain condition have lesser `Priority` than others. Means that you still need to position your units so that they have higher-priority targets in range. Would need to be tested how effective it is and how well it works with a-move, but I think this could work.
- Completely change the iron curtain power to be drain-able instead of "real" invulnerability (Perhaps utilizing `GrantPrerequisiteChargeDrainPower` and `DrainPrerequisitePowerOnDamage`). This solves the problem that autotarget-priorities are an exception but changes the power quite heavily away from how it works in the original.

## Side effects

- Unknown effect on faction balance
- Would either introduce an exception (autotargeting-priorities/temporary invalid targets) or substantial changes to important support power
- Might help with the perception of many players that the chronosphere power is just not good compared to the iron curtain power.

## Alternatives

New area-targeting mode that would order selected actors to only target enemies in that area.

**Comment by pchote** [comment]
The first two proposed solutions can be easily tested/implemented: define an extra conditional `Targetable` trait on `^IronCurtainable` that adds an `Invulnerable` target type to the actor. The `AutoTargetPriority` can then use this in `ValidTargets`/`InvalidTargets`. IMO the second option is the best of those three.

**Comment by MustaphaTR** [comment]
I would like to point out that in Yuri's Revenge (but not in base Red Alert 2) units did not auto-target Iron Curtained enemy units.

**Comment by HappyORA** [comment]
From a competitive players perspective:

There is proposed changes coming after RAGL in tackling the bad design of tanya and the IC (westwoods fault not ORA's). I completely agree with all your points on this post, as well as the first two solutions (even better pchote said they can be added easily). The other proposed changes was reducing the IC time duration.

Making IC units not prioritised is great as it keeps core IC strategies while removing the bad design.

- IC Tanks/TeslaTanks can clear defensive structure lines.
- IC Flaks can still mass murder artillery and air units
- IC V2's can still snipe important structures.
- IC APC/TeslaTanks/Tanks can still crush inf blobs.

And what these changes remove is everything you listed in your post. This is great when coupled with Tanya changes.

The second solution, which pchote thinks is best, would bring back skillfull micro such as old school basepush, placing down one barracks to absorb the initial barrage of a blob to get your own army in a better position to trade. Similar micro would be employed here, low priority IC units targeted to allow the extra time to get an army in better position, as the defending army will waste their first volley on the IC units, and negate defenders advantage (+ the vision the IC units give). I'm not sure myself which is the best of the two solutions, up to debate, but for sure better than what we have currently. It is a lot more skilful to execute than the mindless invulnerable unit charge but I think some players just want that feature gone completely.

overall +1 from me :)

**Comment by pchote** [comment]
The micro implications of the first two options, and a third related option (not the third from above) to help with any debate:

* Option 2 as presented: units will continue to attack if their target is iron curtained, units will automatically target iron curtained units if nothing else is nearby. An explicit stop command is required for the units to target something else.
* Option 1 as presented: units will continue to attack if their target is iron curtained, but go idle if there are no non-curtained targets nearby. Units that aren't already shooting the iron curtained units will therefore automatically acquire other targets that come into range without needing a stop order.
* Set `InvalidTargets` on the weapon instead of `AutoTargetPriority`: units will stop attacking iron curtained units, and cannot target them at all (either automatically or manually). Units will automatically acquire other targets nearby without requiring a stop command.

Basically, targeting priorities would keep the requirement for stop micro retargeting (but remove the broken bit where they retarget the curtained units), versus removing the damage soaking capabilities completely, versus something in between.

**Comment by anjew175** [comment]
Im not against nerfing the IC however I am personally against targetting changes with the IC. The current behaviour is iconic and the expected behaviour from those new and old to CnC (except YR apparently). Also it would be the only exception to the unit targetting system in RA, I can think of, outside of stealth. 

This may be much different in much larger teamgames but I find that for most games the IC is not as overpowered as its made out to be by some. I agree that its very easy to use and how it can feel like "Losing to the computer instead of the enemy's tactic," however there are a few ways to mitigate the effect and have it work towards your favour. This isn't an argument of 'git gud' though, I just disagree that its as strong as is presented. Often its strength is amplified by the advantage someone already has.

Also wrt to 
> "Turns the exisiting defenders advantage into a "Nope, you can't attack here"

This is the side effect of having the vice versa available. One of the primary intents of the IC is to break through the defenders advantage which, as you state, is already very hard. Drastic changes to the IC targetting mechanics may nullify the intent of the superweapon and create more of a stalemate in games (which players already hate from playing Allies vs Allies).


In saying this, option 2 has the least disruption to this behaviour (outside of reducing duration) as it will require the player to intervene if they wish to change their units targetting. This may not be obvious to new players or people of the loop. Obviously people would figure out after they find their IC attacks arent working as well as others but it doesnt seem intiative. At the end of the day though, I don't think these changes will or can address all the issues you mention and I certainly dont think it will change idea that its strong and has the potential to ruin an army. The primary use of IC other than breaking bases or to push any enemy away is to crush the infantry support and a lack of targetting on these units while moving actually makes them easier to kill then if they are stationary. The draining power may address this specific issue though it doesnt address all the issues that have been presented either.

I'd rather see this addressed via balancing rather than shoehorning changes to core mechanics based on anecdotal evidence.

**Comment by Smittytron** [comment]
> Option 2 as presented: units will continue to attack if their target is iron curtained, units will automatically target iron curtained units if nothing else is nearby. An explicit stop command is required for the units to target something else.

This makes the most sense to me. An opponent will still have to micro against an IC supported army. Stop key will be important but won't cover everything as out of range units can't just be attack moved in.

My gut says this will work fine but I expect it to be put through the balance test-map wringer before making it to a PR.

**Comment by MlemandPurrs** [comment]
I would open a clause and state that defenses should keep targeting iron curtained units as its core of IC functionality.
The targeting change makes most sense for units with limited ammo such as Aircraft. 
That being said it shouldnt override an manual attack order given to Aircraft from before the IC was applied as otherwise would make impossible to soak an round of incoming damage from enemy Aircraft with IC'ed flak trucks.

**Comment by tux314159** [comment]
Take a look at 20:45 and 23:00 of this game https://www.youtube.com/watch?v=aVi_EC3BAJw

**Comment by abcdefg30** [comment]
Closing as stale.
